### ISSUE TITLE
Use Factory Pattern for public API

### ISSUE THREAD

**Issue by pogoba** [numtide/deploykit#4]
**Is your feature request related to a problem? Please describe.**

Extending features will explode parameters for many/most functions. 

**Describe the solution you'd like**

Factory pattern: Work on a factory object to hold all settings and configuration variables. This keeps function signatures at more constant length over time. 

**Describe alternatives you've considered**

Insanely long function signatures are needed for all parent functions when adding a new parameter to a low-level function.

**Comment by Mic92** [comment]
You mean argument list passed to `run`, `run_local` etc?

**Comment by pogoba** [comment]
yes. Extending DeployHost with more features will become quite ugly IMHO. 
https://github.com/pogobanane/deploykit/commit/3270cc5af31af645925cb8a85337bb0005ecebb2

**Comment by Mic92** [comment]
Is this possible while maintaining type checks?
